src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnection.cs

```diff
@@ -0,0 +1,578 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Runtime;
+using System.ServiceModel.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal sealed class PipeConnection : IConnection
+    {
+        // common state
+        private readonly NamedPipeClientStream _pipe;
+        private CloseState _closeState;
+        private bool _aborted;
+        private TraceEventType _exceptionEventType;
+        private static readonly byte[] s_zeroBuffer = Array.Empty<byte>();
+
+        // read state
+        private readonly object _readLock = new object();
+        private bool _inReadingState;     // This keeps track of the state machine (IConnection interface).
+        private readonly int _connectionBufferSize;
+        private readonly TaskCompletionSource _atEOFTask;
+        private bool _isAtEOF;
+
+        // write state
+        private readonly object _writeLock = new object();
+        private bool _inWritingState;      // This keeps track of the state machine (IConnection interface).
+        private bool _isShutdownWritten;
+
+        // timeout support
+        private string _timeoutErrorString;
+        private TransferOperation _timeoutErrorTransferOperation;
+
+        public PipeConnection(NamedPipeClientStream namedPipeClient, int connectionBufferSize)
+        {
+            _pipe = namedPipeClient ?? throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+            if (namedPipeClient == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+            if (!namedPipeClient.IsConnected)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+
+            _pipe = namedPipeClient;
+            _closeState = CloseState.Open;
+            _exceptionEventType = TraceEventType.Error;
+            _connectionBufferSize = connectionBufferSize;
+            _atEOFTask = new TaskCompletionSource();
```

Comment:
You probably want this:
```suggestion
            _atEOFTask = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
```
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/0c82fa823796c38d0436c0c08665c944ab12f91d/AsyncGuidance.md#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously

Kind of related, does WCF care about the synchronization context? Should `ConfigureAwait(false)` be added to all awaits?